### PR TITLE
Project-Enumerables

### PR DESCRIPTION
**Pull request by Zubenna**
# Summary of code:
- I implemented my own version of my_each, my_each_with_index, my_select, my_all?, my_any?, my_none?, my_count, my_map, and  my_inject ruby enumerable methods.
- Tested my_inject with the multiply_els method.
- Tested my_map by calling it with a proc.
- Tested all the methods in the file tests.rb

### CODE REVIEW

**Comment by lekegitrepo** [comment]
#### Hi @Zubenna :clap: :clap:
#### Well done, you did a good job so far. However, you need to improve the following methods :point_down: before this project can be approved.

- my_all?
- my_any?
- my_none?
- my_inject

_**They did not output the expected result**_

**_fix these and submit for review_**

### _Reviewed by Temitayo Adeleke._

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### Status: Request changes

**Comment by Zubenna** [comment]
I have modified the methods pointed out. In case there is still a need for modification, Please give a detailed explanation and guide, e.g the supplied argument and expected result so that I will use it in my testing.

File: custom_enumerable_methods.rb
Comment (shshamim63) [inline custom_enumerable_methods.rb]: `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods)) ``
- After calling the given block once for each element in self, returns the array itself
```ruby
[1, 2, 3].my_each {|num| num}  #should return [1, 2, 3]
```

File: custom_enumerable_methods.rb
Comment (shshamim63) [inline custom_enumerable_methods.rb]: `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- After calling the given block once for each element in self, returns the array itself
```ruby
[1, 2, 3].my_each_with_index {|num| num}  #should return [1, 2, 3]
```

File: custom_enumerable_methods.rb
Comment (shshamim63) [inline custom_enumerable_methods.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods)) 
- Just one more case, when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```

File: custom_enumerable_methods.rb
Comment (shshamim63) [inline custom_enumerable_methods.rb]: `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
image
- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
my_proc = Proc.new { |num| num > 10 }
[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :recycle: Required Changes :rocket: 
___
Hi @Zubenna :wave: , almost there. Just need to resolve couple of corner cases and then you will be done with it. Let's get to those cases.
___
### :spades: Mandatory Changes :spades: 
___
- Check out the inline comments below to resolve them.
___
### :clubs: Optional Changes :clubs: 
___
- You should not make any change to the given rubocop configuration. But you have done it. Please check out this [link](https://github.com/microverseinc/linters-config/tree/master/ruby). This is a crucial requirement. But as the previous reviewer forgot it to mention I am adding it in the optional tag.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Comment by Zubenna** [comment]
Requested changes have been implemented. The specified methods have been modified. Required tests and parameters has been implemented in tests.rb file. They returned the expected result.

File: custom_enumerable_methods.rb
Comment (stratospherique) [inline custom_enumerable_methods.rb]: * [ ] my_each returns an Enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each.class # not Enumerator
  ```

File: custom_enumerable_methods.rb
Comment (stratospherique) [inline custom_enumerable_methods.rb]: * [ ] my_all? identical and returns the same thing as ruby's all?. 

- Just one more case, when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```

File: custom_enumerable_methods.rb
Comment (stratospherique) [inline custom_enumerable_methods.rb]: * [ ]  my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
   ```ruby
	my_proc = Proc.new { |num| num > 10 }
	[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
   ```

File: custom_enumerable_methods.rb
Comment (stratospherique) [inline custom_enumerable_methods.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./custom_enumerable_methods.rb:4:in `my_each'
     # ./custom_enumerable_methods.rb:225:in `my_inject'
  ```
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    # NameError:
       undefined local variable or method `length' for 5..50:Range
     # ./custom_enumerable_methods.rb:4:in `my_each'
     # ./custom_enumerable_methods.rb:185:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested :clap: :clap: However, some methods need further refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by Zubenna** [comment]
I have modified the required methods as recommended.
Note that stickler does not accept definition of proc this way => my_proc = Proc.new { |num| num > 10 }
That was why I did not use it that way.

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Great job @Zubenna:clap:. Just a little more step and you can :rocket:**
- See this example, `[1, 2, false].my_all?`. It should return `false` as pointed out in the last review but currently it returns `true`. A similar example was given in the last review but with a `nil` element in the array. The logic is that if there is at least one element that is not a truthy (it could be a false or nil), then my_all? should return false when no block is given.

- Your *my_map* still doesn't handle the case where both a proc and a block are given correctly. Try incorporating this logic in your method:
```
  def my_map(my_proc = nil)
    new_array = []
    if !my_proc
      return to_enum unless block_given?
      my_each { |n| new_array << yield(n) }
    else
      my_each { |n| new_array << my_proc.call(n) }
    end
    new_array
  end
```


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Comment by Zubenna** [comment]
Affected methods have been modified to implement recommendations.

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hello:clap: It seems my_all? was broken while you were trying to implement the last review. Two test cases are now failing:**

:point_right: When a class is passed as argument, it should return true if all the elements are members of that class.
`[1, 2, 3, 4].my_al?l(Integer)` should return `true` 

:point_right: When a regex is passed as argument, it should return true if all the elements match that pattern.
`%w[dog, door, doll].my_all?(/d/)` should return `true`

Please address these cases and submit another code review request.

### Status: Changes Requested.

**Comment by Zubenna** [comment]
Recommended changes has been implemented.

**Review by mikenath223 — CHANGES_REQUESTED** [review]
Hi @Zubenna  👋

You've come quite far on this.🚀

Let's fix the few remaining issues and you are good to go.👌
The issues basically exist with my_all.

- [ ] my_all, when a false block is passed to it, is expected to return false but yet returned true. Kindly fix this.

- [ ] my_all, also when no block or argument is given should return true if the array passed to it has no false member but yours returns false.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

**Comment by Zubenna** [comment]
I have implemented requested change.

**Comment by shshamim63** [comment]
## Status :100: Approved
___
Hi @Zubenna :wave:
Awesome job. You have successfully covered all the changes. and all your methods are working as they should. Nothing is blocking me to approve it. This :+1: is from me.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Pull request merged**
